pallets/loans/src/entities/pricing/external.rs

```diff
@@ -127,42 +150,42 @@ impl<T: Config> ExternalActivePricing<T> {
 
 	pub fn max_borrow_amount(
 		&self,
-		desired_amount: T::Balance,
+		amount: ExternalAmount<T>,
 	) -> Result<T::Balance, DispatchError> {
 		match self.info.max_borrow_amount {
 			MaxBorrowAmount::Quantity(quantity) => {
-				let price = self.current_price()?;
 				let available = quantity.ensure_sub(self.outstanding_quantity)?;
-				Ok(available.ensure_mul_int(price)?)
+				Ok(available.ensure_mul_int(amount.settlement_price)?)
 			}
-			MaxBorrowAmount::NoLimit => Ok(desired_amount),
+			MaxBorrowAmount::NoLimit => Ok(amount.balance()?),
 		}
 	}
 
+	pub fn max_repay_principal(
+		&self,
+		amount: ExternalAmount<T>,
+	) -> Result<T::Balance, DispatchError> {
+		Ok(self
+			.outstanding_quantity
+			.ensure_mul_int(amount.settlement_price)?)
+	}
+
 	pub fn adjust(
 		&mut self,
-		principal_adj: Adjustment<T::Balance>,
+		quantity_adj: Adjustment<T::Rate>,
 		interest: T::Balance,
 	) -> DispatchResult {
-		let quantity_adj = principal_adj.try_map(|principal| -> Result<_, DispatchError> {
-			let price = self.current_price()?;
-
-			let quantity = T::Rate::ensure_from_rational(principal, price)?;
+		self.outstanding_quantity = quantity_adj.ensure_add(self.outstanding_quantity)?;
 
+		let interest_adj = quantity_adj.try_map(|quantity| -> Result<_, DispatchError> {
 			ensure!(
-				quantity.frac().is_zero(),
-				Error::<T>::AmountNotMultipleOfPrice
+				quantity.frac().is_zero() && quantity >= T::Rate::zero(),
```

Comment:
Again, why the change to allow zero

Reply:
It's totally fine to repay with 0 of quantity if you want, for example, to pay only interest.